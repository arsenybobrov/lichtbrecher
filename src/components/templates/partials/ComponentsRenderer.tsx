import React from 'react';
import get from 'lodash/get';
import uniqueId from 'lodash/uniqueId';
import SLICES_MAP from '../../../../prismic/slices';
import CmsElmtWrapper from '../../organisms/CmsElmtWrapper/CmsElmtWrapper';

interface ComponentsRendererProps {
  slices: [
    {
      slice_type: string;
      items: any;
      primary: any;
    }
  ];
}

interface UndefinedSliceProps {
  sliceName: string;
}

const UndefinedSlice: React.FC<UndefinedSliceProps> = ({ sliceName, ...rest }) => (
  <>
    <div>&quot;{sliceName}&quot; is not defined yet.</div>
    <div>Data: {JSON.stringify(rest, null, 2)}</div>
  </>
);

const ComponentsRenderer: React.FC<ComponentsRendererProps> = ({ slices }) => (
  <>
    {
      slices.map((slice, idx) => {
        return (
          <CmsElmtWrapper
            key={uniqueId()}
            id={slice.primary.anchor_id ?? undefined}
            sliceName={slice.slice_type}
            nextSlice={slices[idx + 1]?.slice_type ?? undefined}
          >
            {React.createElement(
              get(SLICES_MAP, slice.slice_type, UndefinedSlice),
              {
                primary: slice.primary,
                items: slice.items,
                sliceIndex: idx,
                sliceName: slice.slice_type,
              },
              null
            )}
          </CmsElmtWrapper>
        )
      })
    }
  </>
);

export default ComponentsRenderer;
