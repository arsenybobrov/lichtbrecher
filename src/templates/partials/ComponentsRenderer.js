import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import uniqueId from 'lodash/uniqueId';
import SLICE_MAP from '../../../api/prismic/slices/map';

const ComponentsRenderer = ({
  slices,
}) => (
  <>
    {
      slices.map((slice) => (
        <div key={uniqueId()}>
          {React.createElement(
            get(SLICE_MAP, slice.slice_type, SLICE_MAP.dummy),
            { primary: slice.primary, items: slice.items, sliceName: slice.slice_type },
            null
          )}
        </div>
      ))
    }
  </>
);

ComponentsRenderer.propTypes = {
  slices: PropTypes.array,
};

export default ComponentsRenderer;
