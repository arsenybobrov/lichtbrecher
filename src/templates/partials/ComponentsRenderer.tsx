import React from 'react';
import get from 'lodash/get';
import uniqueId from 'lodash/uniqueId';
import SLICE_MAP from '../../../api/prismic/slices/map';

interface Props {
  slices: [
    {
      slice_type: string;
      items: [];
      primary: {};
    }
  ];
}

const ComponentsRenderer: React.FC<Props> = ({
  slices,
}) => (
  <>
    {
      slices.map((slice) => (
        <div key={uniqueId()}>
          {React.createElement(
            get(SLICE_MAP, slice.slice_type, SLICE_MAP.dummy),
            { primary: slice.primary, items: slice.items, sliceName: slice.slice_type },
            null
          )}
        </div>
      ))
    }
  </>
);

export default ComponentsRenderer;
