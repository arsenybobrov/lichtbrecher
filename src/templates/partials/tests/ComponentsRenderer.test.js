import React from 'react';
import renderer from 'react-test-renderer';
import ComponentsRenderer from '../ComponentsRenderer';

const DataUnknownSlicesMock = {
  slices: [
    {
      slice_type: 'slice_name',
    },
    {
      slice_type: 'slice_another_name',
    },
  ],
};

describe('ComponentRenderer snapshot', () => {
  it.each`
    name                                                                                           |  mock
    ${'renders 2 dummy components, because slice_name & slice_another_name is not inside map.js'}  |  ${DataUnknownSlicesMock}
  `('$name', ({ mock }) => {
  const tree = renderer
    .create(
      <ComponentsRenderer {...mock} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
});
