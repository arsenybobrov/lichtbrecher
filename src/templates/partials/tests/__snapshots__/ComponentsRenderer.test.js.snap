// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextLink snapshot renders 2 dummy components, because slice_name & slice_another_name is not inside map.js 1`] = `
Array [
  <div>
    <div
      className="DummySlice__Wrapper-sc-1p1rpu-0 jrtuih"
    >
      the slice 
      slice_name
       does not exist yet...
    </div>
  </div>,
  <div>
    <div
      className="DummySlice__Wrapper-sc-1p1rpu-0 jrtuih"
    >
      the slice 
      slice_another_name
       does not exist yet...
    </div>
  </div>,
]
`;
